### ISSUE TITLE
[JENKINS-13153] "Upgrade Automatically" does not seem to work on Windows

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#12542]
<p>After moving to a Jenkins installation that runs as a service. I noticed it now had the option to download new versions of itself and upgrade. I tried that with 1.452 and was a bit confused that it restarted but still showed 1.452 as version number. But also gives me the option to downgrade to 1.452.</p>

<p>After a couple more tries I think I can conclude that what happens is that it downloads the new version just fine (as jenkins.war.tmp) and copies the current version fine as well (jenkins.war.bak). But it never actually gets around to replacing the current running version with the new one.</p>

<p>This appears to be the relevant part of the jenkins.err.log:</p>

```
INFO: Starting the installation of jenkins.war on behalf of kpe
Mar 19, 2012 9:17:19 AM hudson.model.UpdateCenter$UpdateCenterConfiguration download
INFO: Downloading jenkins.war
Mar 19, 2012 9:17:19 AM hudson.model.UpdateCenter doSafeRestart
INFO: Scheduling Jenkins reboot
Mar 19, 2012 9:17:40 AM hudson.model.UpdateCenter$DownloadJob run
INFO: Installation successful: jenkins.war

```

<p>But I am a bit confused on how this would ever actually work on Windows? Since files are generally locked up pretty tight. So replacing or overwriting a file that Java is actively using would be practically impossible. And instead it would require some sort of utility running before starting Jenkins to replace the war file with the newly downloaded one.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kristoffer.jpg" title="kristoffer's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kristoffer">kristoffer</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153" target="_blank">"Upgrade Automatically" does not seem to work on Windows</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=hp">hp</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-09-15T19:59:38+00:00
<li><b>votes</b>: 4
<li><b>watchers</b>: 10
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>After moving to a Jenkins installation that runs as a service. I noticed it now had the option to download new versions of itself and upgrade. I tried that with 1.452 and was a bit confused that it restarted but still showed 1.452 as version number. But also gives me the option to downgrade to 1.452.</p>

<p>After a couple more tries I think I can conclude that what happens is that it downloads the new version just fine (as jenkins.war.tmp) and copies the current version fine as well (jenkins.war.bak). But it never actually gets around to replacing the current running version with the new one.</p>

<p>This appears to be the relevant part of the jenkins.err.log:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>INFO: Starting the installation of jenkins.war on behalf of kpe
Mar 19, 2012 9:17:19 AM hudson.model.UpdateCenter$UpdateCenterConfiguration download
INFO: Downloading jenkins.war
Mar 19, 2012 9:17:19 AM hudson.model.UpdateCenter doSafeRestart
INFO: Scheduling Jenkins reboot
Mar 19, 2012 9:17:40 AM hudson.model.UpdateCenter$DownloadJob run
INFO: Installation successful: jenkins.war
</pre>
</div></div>

<p>But I am a bit confused on how this would ever actually work on Windows? Since files are generally locked up pretty tight. So replacing or overwriting a file that Java is actively using would be practically impossible. And instead it would require some sort of utility running before starting Jenkins to replace the war file with the newly downloaded one.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Windows 2003 R3<br/>
Java 1.6.0_26<br/>
Apache httpd 2.2 as frontend (apj13 proxy)
```
</details>
<details><summary><i>2 attachments</i></summary>

- [Screen Shot 2016-06-08 at 2.34.00 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/32943/Screen%20Shot%202016-06-08%20at%202.34.00%20PM.png)
  > ![Screen Shot 2016-06-08 at 2.34.00 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/32943/Screen%20Shot%202016-06-08%20at%202.34.00%20PM.png)
- [jenkins.xml](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/32942/jenkins.xml)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [https://github.com/jenkinsci/jenkins/pull/2992](https://github.com/jenkinsci/jenkins/pull/2992)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tang.png" title="tang's avatar" /> tang:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=175287&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-175287">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have the same problem, and the most strange thing is that there exists no exception or warning during installation.</p>

<p>Maybe <tt>jenkins.exe</tt> is a good place for doing this?</p>
</pre>
</details>

</ul>
</details>

<p>I have the same problem, and the most strange thing is that there exists no exception or warning during installation.</p>

<p>Maybe <tt>jenkins.exe</tt> is a good place for doing this?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dhs:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=178504&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-178504">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Maybe this functionality should generaly be disabled on Windows?</p>
</pre>
</details>

</ul>
</details>

<p>Maybe this functionality should generaly be disabled on Windows?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212065&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212065">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is this still a problem in recent Jenkins versions?</p>
</pre>
</details>

</ul>
</details>

<p>Is this still a problem in recent Jenkins versions?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kristoffer.jpg" title="kristoffer's avatar" /> kristoffer:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212211&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212211">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Not a clue. I no longer work at the place where we had that installation. And don't really have anywhere to replicate the setup either.</p>
</pre>
</details>

</ul>
</details>

<p>Not a clue. I no longer work at the place where we had that installation. And don't really have anywhere to replicate the setup either.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212422&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212422">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for the update Kristoffer!</p>

<p>Resolving as incomplete as no information about the issue in recent Jenkins versions is known. (there were a few improvements to jenkins.exe since this was filed, so I have reason to believe this may have been fixed) </p>

<p>If this happens in a recent (no older than six weeks) version of Jenkins, please just file a new issue. Thanks!</p>
</pre>
</details>

</ul>
</details>

<p>Thanks for the update Kristoffer!</p>

<p>Resolving as incomplete as no information about the issue in recent Jenkins versions is known. (there were a few improvements to jenkins.exe since this was filed, so I have reason to believe this may have been fixed) </p>

<p>If this happens in a recent (no older than six weeks) version of Jenkins, please just file a new issue. Thanks!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/phancox.jpg" title="phancox's avatar" /> phancox:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212451&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212451">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is still a problem as of version 1.574</p>

<p>Attempted automatic upgrade from 1.574 to 1.582 and screen stuck on "Jenkins restarting ..."  Stop the Jenkins service and copy "jenkins.war.tmp" to "jenkins.war" and restart service.  Upgrade is now complete.</p>

<p>Appears as if upgrade is waiting for the downloaded temporary war to replace the currently active war and this is hanging.  My Jenkins is installed in the default "C:\Program Files (x86)\Jenkins\" directory  so I would expect that the update process would need elevated privileges, though I suspect it already has them since it can place the new downloaded war file in that directory.</p>

<p>Hope that helps.</p>
</pre>
</details>

</ul>
</details>

<p>This is still a problem as of version 1.574</p>

<p>Attempted automatic upgrade from 1.574 to 1.582 and screen stuck on "Jenkins restarting ..."  Stop the Jenkins service and copy "jenkins.war.tmp" to "jenkins.war" and restart service.  Upgrade is now complete.</p>

<p>Appears as if upgrade is waiting for the downloaded temporary war to replace the currently active war and this is hanging.  My Jenkins is installed in the default "C:\Program Files (x86)\Jenkins\" directory  so I would expect that the update process would need elevated privileges, though I suspect it already has them since it can place the new downloaded war file in that directory.</p>

<p>Hope that helps.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212454&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212454">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Peter: Do you remember the Jenkins version you originally installed, and could look up the creation + last modification dates of jenkins.exe? AFAIK there is a (now resolved) bug that prevented that from getting updated on Jenkins update, and one of the fixed issues was a failure to restart Jenkins as well. So this appears to be an outdated jenkins.exe.</p>

<p>Current jenkins.exe should have a date from early April, assuming it doesn't get changed when packaged into Jenkins.</p>
</pre>
</details>

</ul>
</details>

<p>Peter: Do you remember the Jenkins version you originally installed, and could look up the creation + last modification dates of jenkins.exe? AFAIK there is a (now resolved) bug that prevented that from getting updated on Jenkins update, and one of the fixed issues was a failure to restart Jenkins as well. So this appears to be an outdated jenkins.exe.</p>

<p>Current jenkins.exe should have a date from early April, assuming it doesn't get changed when packaged into Jenkins.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/phancox.jpg" title="phancox's avatar" /> phancox:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212458&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212458">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>My jenkins.exe is dated 24-Mar-2014.</p>

<p>Should I uninstall/install the latestWindows package?</p>
</pre>
</details>

</ul>
</details>

<p>My jenkins.exe is dated 24-Mar-2014.</p>

<p>Should I uninstall/install the latestWindows package?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212461&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212461">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Not sure how to best do this. Possibly through reinstall. IIRC a user reported he just replaced the jenkins.exe, so you could also try the following:</p>

<p>If you find a file identical to your jenkins.exe among the exe files here (as sanity check; I'm not 100% sure this is the correct file): <a href="http://repo.jenkins-ci.org/releases/com/sun/winsw/winsw/" class="external-link" target="_blank" rel="nofollow noopener">http://repo.jenkins-ci.org/releases/com/sun/winsw/winsw/</a>
... then just replace it with the exe in the 1.16 folder.</p>
</pre>
</details>

</ul>
</details>

<p>Not sure how to best do this. Possibly through reinstall. IIRC a user reported he just replaced the jenkins.exe, so you could also try the following:</p>

<p>If you find a file identical to your jenkins.exe among the exe files here (as sanity check; I'm not 100% sure this is the correct file): <a href="http://repo.jenkins-ci.org/releases/com/sun/winsw/winsw/" class="external-link" target="_blank" rel="nofollow noopener">http://repo.jenkins-ci.org/releases/com/sun/winsw/winsw/</a><br/>
... then just replace it with the exe in the 1.16 folder.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/phancox.jpg" title="phancox's avatar" /> phancox:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212468&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212468">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Didn't uninstall. Just installed latest package over existing installation and reapplied changes I'd made to "jenkins.xml".</p>

<p>So far so good.  Have to wait and see if next auto update works.</p>

<p>THANKS</p>
</pre>
</details>

</ul>
</details>

<p>Didn't uninstall. Just installed latest package over existing installation and reapplied changes I'd made to "jenkins.xml".</p>

<p>So far so good.  Have to wait and see if next auto update works.</p>

<p>THANKS</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/phancox.jpg" title="phancox's avatar" /> phancox:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212838&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212838">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Still no luck unfortunately.  Attempted automatic upgrade from 1.582 to 1.583 and experienced exactly the same behaviour as previously.</p>

<p>REGARDS
Peter</p>
</pre>
</details>

</ul>
</details>

<p>Still no luck unfortunately.  Attempted automatic upgrade from 1.582 to 1.583 and experienced exactly the same behaviour as previously.</p>

<p>REGARDS<br/>
Peter</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=212839&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212839">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Comments indicate this is still relevant.</p>
</pre>
</details>

</ul>
</details>

<p>Comments indicate this is still relevant.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mcsf:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=214043&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-214043">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>On Windows, did an automatic upgrade from 1.554.2 to 1.565.3, and in the jenkins.err.log I see this:</p>

<p>INFO: Scheduled a replacement of jenkins.exe</p>

<p>Although when Jenkins comes up, the message at the bottom indicates that 1.565.3 is running.</p>

<p>Instead of checking the checkbox 'Restart Jenkins when installation is complete and no jobs are running', we always restart Jenkins via the Services control panel, after the downloads are complete.</p>

<p>I ended up updating the file manually, then restarting Jenkins.
After doing so, the 'INFO: Scheduled a replacement of jenkins.exe' message went away.
It is kind of disturbing to see that jenkins.exe was never actually updated since 05/02/2014 until today, which is when I first noticed that message in the Jenkins log.</p>

<p>05/02/2014  09:15 AM            50,624 jenkins.exe.old
10/15/2014  06:18 PM            55,776 jenkins.exe.new
10/15/2014  06:18 PM            55,776 jenkins.exe</p>
</pre>
</details>

</ul>
</details>

<p>On Windows, did an automatic upgrade from 1.554.2 to 1.565.3, and in the jenkins.err.log I see this:</p>

<p>INFO: Scheduled a replacement of jenkins.exe</p>

<p>Although when Jenkins comes up, the message at the bottom indicates that 1.565.3 is running.</p>

<p>Instead of checking the checkbox 'Restart Jenkins when installation is complete and no jobs are running', we always restart Jenkins via the Services control panel, after the downloads are complete.</p>

<p>I ended up updating the file manually, then restarting Jenkins.<br/>
After doing so, the 'INFO: Scheduled a replacement of jenkins.exe' message went away.<br/>
It is kind of disturbing to see that jenkins.exe was never actually updated since 05/02/2014 until today, which is when I first noticed that message in the Jenkins log.</p>

<p>05/02/2014  09:15 AM    50,624 jenkins.exe.old<br/>
10/15/2014  06:18 PM    55,776 jenkins.exe.new<br/>
10/15/2014  06:18 PM    55,776 jenkins.exe</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=255537&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-255537">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Anyone still experiencing this on recent Jenkins versions, with up to date jenkins.exe?</p>
</pre>
</details>

</ul>
</details>

<p>Anyone still experiencing this on recent Jenkins versions, with up to date jenkins.exe?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ganthore.png" title="ganthore's avatar" /> ganthore:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=260477&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-260477">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hey Daniel,</p>

<p>I've managed to reproduce this issue multiple times using a full admin windows account while using the latest copy of winsw (1.18). It seems to me that the process that writes jenkins.war.tmp fails silently. I can confirm that the correct update is being downloaded, stored as jenkins.war.tmp and making a copy of the old file as jenkins.war.bak.</p>

<p><span class="image-wrap" style=""><a id="32943_thumb" href="https://issues.jenkins.io/secure/attachment/32943/32943_Screen+Shot+2016-06-08+at+2.34.00+PM.png" title="Screen Shot 2016-06-08 at 2.34.00 PM.png" file-preview-type="image" file-preview-id="32943" file-preview-title="Screen Shot 2016-06-08 at 2.34.00 PM.png"><img src="https://issues.jenkins.io/secure/thumbnail/32943/_thumb_32943.png" style="border: 0px solid black" role="presentation"/></a></span></p>

<p>I suspect it's happening due to the jenkins.exe still holding a file handle on the original war file and fails silently based off how I understand what's going on in the core code:
<a href="https://github.com/jenkinsci/jenkins/blob/cf64ba04fe8b0312bb4492766d62b5a00b790432/core/src/main/java/hudson/lifecycle/Lifecycle.java#L137-L170" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/cf64ba04fe8b0312bb4492766d62b5a00b790432/core/src/main/java/hudson/lifecycle/Lifecycle.java#L137-L170</a></p>

<p>It may be helpful to have the</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">canRewriteHudsonWar()</pre>
</div></div>
<p>function report success vs failure in the logger. I guess it would be best to avoid throwing an exception to help prevent introducing API breakage.</p>

<p>The windows event viewer will throw the following details when the service is terminated (that is, when you check the restart box), but even the OS logs fail to report a problem, so this fails silently by all accounts right now. No additional errors are reported in the wrapper log. Here is the windows event at the time the JVM is restarted after the update is downloaded to the system:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Log Name:      Application
Source:        Jenkins
Date:          6/8/2016 2:12:14 PM
Event ID:      0
Task Category: None
Level:         Information
Keywords:      Classic
User:          N/A
Computer:      reubenWin
Description:
Child process [4860 - C:\jenkins\java\jre1.8.0_92\bin\java -server -d64 -Xrs -Xms2048m -Xmx2048m -XX:+AlwaysPreTouch -Dhudson.lifecycle=hudson.lifecycle.WindowsServiceLifecycle -Djava.awt.headless=<span class="code-keyword">true</span> -Djava.net.preferIPv4Stack=<span class="code-keyword">true</span> -Djenkins.model.Jenkins.logStartupPerformance=<span class="code-keyword">true</span> -Dhudson.udp=-1 -Dhudson.DNSMultiCast.disabled=<span class="code-keyword">true</span> -jar <span class="code-quote">"C:\jenkins\jenkins.war"</span> --httpListenAddress=0.0.0.0 --httpsPort=-1 --httpPort=8080 --ajp13Port=-1 --webroot=C:\jenkins\webroot] terminated with -1073741510
Event Xml:
&lt;Event xmlns=<span class="code-quote">"http:<span class="code-comment">//schemas.microsoft.com/win/2004/08/events/event"</span>&gt;
</span>  &lt;<span class="code-object">System</span>&gt;
    &lt;Provider Name=<span class="code-quote">"Jenkins"</span> /&gt;
    &lt;EventID Qualifiers=<span class="code-quote">"0"</span>&gt;0&lt;/EventID&gt;
    &lt;Level&gt;4&lt;/Level&gt;
    &lt;Task&gt;0&lt;/Task&gt;
    &lt;Keywords&gt;0x80000000000000&lt;/Keywords&gt;
    &lt;TimeCreated SystemTime=<span class="code-quote">"2016-06-08T18:12:14.000000000Z"</span> /&gt;
    &lt;EventRecordID&gt;1159&lt;/EventRecordID&gt;
    &lt;Channel&gt;Application&lt;/Channel&gt;
    &lt;Computer&gt;reubenWin&lt;/Computer&gt;
    &lt;Security /&gt;
  &lt;/<span class="code-object">System</span>&gt;
  &lt;EventData&gt;
    &lt;Data&gt;Child process [4860 - C:\jenkins\java\jre1.8.0_92\bin\java -server -d64 -Xrs -Xms2048m -Xmx2048m -XX:+AlwaysPreTouch -Dhudson.lifecycle=hudson.lifecycle.WindowsServiceLifecycle -Djava.awt.headless=<span class="code-keyword">true</span> -Djava.net.preferIPv4Stack=<span class="code-keyword">true</span> -Djenkins.model.Jenkins.logStartupPerformance=<span class="code-keyword">true</span> -Dhudson.udp=-1 -Dhudson.DNSMultiCast.disabled=<span class="code-keyword">true</span> -jar <span class="code-quote">"C:\jenkins\jenkins.war"</span> --httpListenAddress=0.0.0.0 --httpsPort=-1 --httpPort=8080 --ajp13Port=-1 --webroot=C:\jenkins\webroot] terminated with -1073741510&lt;/Data&gt;
  &lt;/EventData&gt;
&lt;/Event&gt;
</pre>
</div></div>

<p>I've attached a copy of my jenkins.xml file if that helps:
<span class="nobr"><a href="https://issues.jenkins.io/secure/attachment/32942/32942_jenkins.xml" title="jenkins.xml attached to JENKINS-13153">jenkins.xml<sup><img class="rendericon" src="https://issues.jenkins.io/images/icons/link_attachment_7.gif" height="7" width="7" align="absmiddle" alt="" border="0"/></sup></a></span></p>

<p>Note that I intentionally use custom folder paths to help prevent windows from locking up files in use, hence the separation of individual directories like this:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">c:\jenkins
c:\jenkins\webroot
c:\jenkins\tmp
c:\jenkins\data
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Hey Daniel,</p>

<p>I've managed to reproduce this issue multiple times using a full admin windows account while using the latest copy of winsw (1.18). It seems to me that the process that writes jenkins.war.tmp fails silently. I can confirm that the correct update is being downloaded, stored as jenkins.war.tmp and making a copy of the old file as jenkins.war.bak.</p>

<p><span class="image-wrap" style=""><a id="32943_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/32943/Screen%20Shot%202016-06-08%20at%202.34.00%20PM.png" title="Screen Shot 2016-06-08 at 2.34.00 PM.png" file-preview-type="image" file-preview-id="32943" file-preview-title="Screen Shot 2016-06-08 at 2.34.00 PM.png"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/32943/Screen%20Shot%202016-06-08%20at%202.34.00%20PM.png" style="border: 0px solid black" role="presentation"/></a></span></p>

<p>I suspect it's happening due to the jenkins.exe still holding a file handle on the original war file and fails silently based off how I understand what's going on in the core code:<br/>
<a href="https://github.com/jenkinsci/jenkins/blob/cf64ba04fe8b0312bb4492766d62b5a00b790432/core/src/main/java/hudson/lifecycle/Lifecycle.java#L137-L170" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/cf64ba04fe8b0312bb4492766d62b5a00b790432/core/src/main/java/hudson/lifecycle/Lifecycle.java#L137-L170</a></p>

<p>It may be helpful to have the</p>

<pre>
canRewriteHudsonWar()</pre>
<p>function report success vs failure in the logger. I guess it would be best to avoid throwing an exception to help prevent introducing API breakage.</p>

<p>The windows event viewer will throw the following details when the service is terminated (that is, when you check the restart box), but even the OS logs fail to report a problem, so this fails silently by all accounts right now. No additional errors are reported in the wrapper log. Here is the windows event at the time the JVM is restarted after the update is downloaded to the system:</p>


<pre>
Log Name:      Application
Source:Jenkins
Date:  6/8/2016 2:12:14 PM
Event ID:      0
Task Category: None
Level: Information
Keywords:      Classic
User:  N/A
Computer:      reubenWin
Description:
Child process [4860 - C:\jenkins\java\jre1.8.0_92\bin\java -server -d64 -Xrs -Xms2048m -Xmx2048m -XX:+AlwaysPreTouch -Dhudson.lifecycle=hudson.lifecycle.WindowsServiceLifecycle -Djava.awt.headless=<span class="code-keyword">true</span> -Djava.net.preferIPv4Stack=<span class="code-keyword">true</span> -Djenkins.model.Jenkins.logStartupPerformance=<span class="code-keyword">true</span> -Dhudson.udp=-1 -Dhudson.DNSMultiCast.disabled=<span class="code-keyword">true</span> -jar <span class="code-quote">"C:\jenkins\jenkins.war"</span> --httpListenAddress=0.0.0.0 --httpsPort=-1 --httpPort=8080 --ajp13Port=-1 --webroot=C:\jenkins\webroot] terminated with -1073741510
Event Xml:
<Event xmlns=<span class="code-quote">"http:<span class="code-comment">//schemas.microsoft.com/win/2004/08/events/event"</span>>
</span>  <<span class="code-object">System</span>>
    <Provider Name=<span class="code-quote">"Jenkins"</span> />
    <EventID Qualifiers=<span class="code-quote">"0"</span>>0</EventID>
    <Level>4</Level>
    <Task>0</Task>
    <Keywords>0x80000000000000</Keywords>
    <TimeCreated SystemTime=<span class="code-quote">"2016-06-08T18:12:14.000000000Z"</span> />
    <EventRecordID>1159</EventRecordID>
    <Channel>Application</Channel>
    <Computer>reubenWin</Computer>
    <Security />
  </<span class="code-object">System</span>>
  <EventData>
    <Data>Child process [4860 - C:\jenkins\java\jre1.8.0_92\bin\java -server -d64 -Xrs -Xms2048m -Xmx2048m -XX:+AlwaysPreTouch -Dhudson.lifecycle=hudson.lifecycle.WindowsServiceLifecycle -Djava.awt.headless=<span class="code-keyword">true</span> -Djava.net.preferIPv4Stack=<span class="code-keyword">true</span> -Djenkins.model.Jenkins.logStartupPerformance=<span class="code-keyword">true</span> -Dhudson.udp=-1 -Dhudson.DNSMultiCast.disabled=<span class="code-keyword">true</span> -jar <span class="code-quote">"C:\jenkins\jenkins.war"</span> --httpListenAddress=0.0.0.0 --httpsPort=-1 --httpPort=8080 --ajp13Port=-1 --webroot=C:\jenkins\webroot] terminated with -1073741510</Data>
  </EventData>
</Event>
</pre>

<p>I've attached a copy of my jenkins.xml file if that helps:<br/>
<span class="nobr"><a href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/32942/jenkins.xml" title="jenkins.xml attached to JENKINS-13153">jenkins.xml<sup><img class="rendericon" src="https://issues.jenkins.io/images/icons/link_attachment_7.gif" height="7" width="7" align="absmiddle" alt="" border="0"/></sup></a></span></p>

<p>Note that I intentionally use custom folder paths to help prevent windows from locking up files in use, hence the separation of individual directories like this:</p>


<pre>
c:\jenkins
c:\jenkins\webroot
c:\jenkins\tmp
c:\jenkins\data
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hp:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=311211&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-311211">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It seems that this happens when when running with a custom JENKINS_HOME. If, so the "jenkins.copies" file is written to JENKINS_HOME and will not be used by the service wrapper when restarting. I have created a pull-request [#2992|<a href="https://github.com/jenkinsci/jenkins/pull/2992" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/2992</a>] with a proposed solution. Using environment variable BASE which is set by the service wrapper. </p>
</pre>
</details>

</ul>
</details>

<p>It seems that this happens when when running with a custom JENKINS_HOME. If, so the "jenkins.copies" file is written to JENKINS_HOME and will not be used by the service wrapper when restarting. I have created a pull-request [#2992|<a href="https://github.com/jenkinsci/jenkins/pull/2992" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/2992</a>] with a proposed solution. Using environment variable BASE which is set by the service wrapper. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=312515&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-312515">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: hplatou
Path:
 core/src/main/java/hudson/lifecycle/WindowsServiceLifecycle.java
<a href="http://jenkins-ci.org/commit/jenkins/0efdf8fb4f8c56f1f32fb390c472cb2e98e67f56" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0efdf8fb4f8c56f1f32fb390c472cb2e98e67f56</a>
Log:
  <a href="https://issues.jenkins.io/browse/JENKINS-13153" title="&quot;Upgrade Automatically&quot; does not seem to work on Windows" class="issue-link" data-issue-key="JENKINS-13153"><del>JENKINS-13153</del></a> - Use directory from env:BASE when writing jenkins.copies (#2992)</p>

<p><a href="https://issues.jenkins.io/browse/JENKINS-13153" title="&quot;Upgrade Automatically&quot; does not seem to work on Windows" class="issue-link" data-issue-key="JENKINS-13153"><del>JENKINS-13153</del></a> - Use directory from env:BASE when writing jenkins.copies</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: hplatou<br/>
Path:<br/>
 core/src/main/java/hudson/lifecycle/WindowsServiceLifecycle.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/0efdf8fb4f8c56f1f32fb390c472cb2e98e67f56" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0efdf8fb4f8c56f1f32fb390c472cb2e98e67f56</a><br/>
Log:<br/>
  <a href="https://issues.jenkins.io/browse/JENKINS-13153" title=""Upgrade Automatically" does not seem to work on Windows" class="issue-link" data-issue-key="JENKINS-13153"><del>JENKINS-13153</del></a> - Use directory from env:BASE when writing jenkins.copies (#2992)</p>

<p><a href="https://issues.jenkins.io/browse/JENKINS-13153" title=""Upgrade Automatically" does not seem to work on Windows" class="issue-link" data-issue-key="JENKINS-13153"><del>JENKINS-13153</del></a> - Use directory from env:BASE when writing jenkins.copies</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-13153?focusedId=313941&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-313941">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Resolved in 2.77 (I think).</p>
</pre>
</details>

</ul>
</details>

<p>Resolved in 2.77 (I think).</p>
